src/v/pandaproxy/json/iobuf.h

```diff
@@ -94,16 +93,44 @@ class rjson_serialize_impl<iobuf> {
         if (buf.empty()) {
             return w.Null();
         }
-        iobuf_parser p{std::move(buf)};
-        auto str = p.read_string(p.bytes_left());
-        static_assert(str.padding(), "StringStream requires null termination");
+        ::json::chunked_input_stream ss{std::move(buf)};
         ::json::Reader reader;
-        ::json::StringStream ss{str.c_str()};
         return reader.Parse(ss, w);
     };
 
 private:
     serialization_format _fmt;
 };
 
+///\brief Wrap a std::istream around an iobuf
+class iobuf_istream {
+public:
+    explicit iobuf_istream(iobuf&& buf)
```

Comment:
by value?

Reply:
> by value?

It can't be copied; the decision to copy, share, or move must be done at the call site regardless, so I saw no need.

Reply:
the benefit is value semantics. iobuf is designed to be very cheap to move.

Reply:
I'm not sure I understand what you're asking. The interface currently requires:
```c++
iobuf buf;
// either of the following:
iobuf_istream is{std::move(buf)};
iobuf_istream is{buf.copy()};
iobuf_istream is{buf.share(0, buf.size_bytes()()};
```
Accepting it by value doesn't change that, right?

Reply:
all i'm saying is that when passing by value, there are a lot of additional properties that can be implied just by seeing that it is passed by value.

when accepting an r-value reference, it's important to go review the code to see if the function accepting it actually takes ownership in all code paths, otherwise the caller can be left with a not-moved-from value (even if it passed with std::move), which has been a footgun in the past. it's clearly correct here, fwiw.

so i tend to ask that we reserve r-value references for places where the performance really matters. maybe it matters here? not a big deal, but I think we should be making that trade off in development. this is another reason why it would be great to see these utilities be broken out into separate PRs. right here in the middle of a big PR it's all very cloudy.

Reply:
Fair point. The arg isn't copyable, not taken by forwarding reference, and not futurized, so I don't think it applies, but it's not a general case. We should be able to provide generalised advice, or even markup (I'm thinking gal ownership) to advertise intent.

Reply:
the footgun still applies, right?

```
f(iobuf&&) {}
iobuf b("abc");
f(std::move(b));
assert b == "abc" will be true
```

Reply:
>We should be able to provide generalised advice, or even markup (I'm thinking gsl ownership) to advertise intent.

doesn't pass-by-value clearly advertise?

Reply:
> the footgun still applies, right?
> 
> ```
> f(iobuf&&) {}
> iobuf b("abc");
> f(std::move(b));
> assert b == "abc" will be true
> ```

No, I don't think so. It's a sink parameter that's always moved.

Reply:
> > We should be able to provide generalised advice, or even markup (I'm thinking gsl ownership) to advertise intent.
> 
> doesn't pass-by-value clearly advertise?

Yes, but it may be a pessimisation. I see no benefit here.

Reply:
Changed